chat-android/src/test/java/com/ably/chat/TaskSchedulerTest.kt

```diff
@@ -0,0 +1,57 @@
+package com.ably.chat
+
+import java.util.concurrent.Executors
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.test.runTest
+import org.junit.Assert
+import org.junit.Test
+
+class TaskSchedulerTest {
+
+    @Test
+    fun `should perform given operation`() = runTest {
+        val singleThreadedDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+        val taskScheduler = TaskScheduler(CoroutineScope(singleThreadedDispatcher))
+        val taskResult = taskScheduler.schedule {
+            delay(3000)
+            return@schedule "Operation Success!"
+        }
+        val result = taskResult.await()
+        Assert.assertTrue(result.isSuccess)
+        Assert.assertFalse(result.isFailure)
+        Assert.assertEquals("Operation Success!", result.getOrNull())
+    }
+
+    @Test
+    fun `should perform mutually exclusive operations with given priority`() = runTest {
+        val singleThreadedDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+        val taskScheduler = TaskScheduler(CoroutineScope(singleThreadedDispatcher))
+        val taskResults = mutableListOf<TaskResult<Int>>()
+        var operationInProgress = false
+        var counter = 0
+
+        repeat(20) {
+            val result = taskScheduler.schedule(it) {
+                if (operationInProgress) {
+                    throw IllegalStateException("Can't perform operation when other operation is going on")
+                }
+                operationInProgress = true
+                delay((200..800).random().toDuration(DurationUnit.MILLISECONDS))
```

Comment:
I believe this code does nothing, because all delays inside `runTest` are skipped, [see](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-test.html)

Reply:
I agree, `runTest` skips use of `delay`. 
But, that's only true when code is run in the context of a given test coroutineScope. 
Since we are passing custom `CoroutineScope`, this doesn't skip delay. 
You might like to run the test on the local by changing values to say 
`delay((2000..2100).random().toDuration(DurationUnit.MILLISECONDS))`

Reply:
Test will take more time to run

Reply:
missed it, you are right! But nevertheless do we really need any of this delays? Not sure delays make tests more reliable